### PR TITLE
fix(iroh-net): improve hole punching when relying on call-me-maybe

### PR DESCRIPTION
**Pull request by dignifiedquire**
- only consider active endpoints for call-me-maybe checks
- insert call me maybe addrs into the peer map

### CODE REVIEW

File: iroh-net/src/magic_endpoint.rs [hunk 1]
```diff
@@ -673,7 +672,8 @@ mod tests {
                     let mut buf = [0u8, 5];
                     stream.read_exact(&mut buf).await.unwrap();
                     info!("Accepted 1 stream, received {buf:?}.  Closing now.");
-                    ep.close(7u8.into(), b"bye").await.unwrap();
+                    // close the stream
+                    conn.close(7u8.into(), b"bye");
```
Comment (flub) [inline iroh-net/src/magic_endpoint.rs:676]: by not awaiting this you're never closing this?

Comment (dignifiedquire) [reply iroh-net/src/magic_endpoint.rs:676]: there is no way to actually wait for a connection close asfaik

File: iroh-net/src/magicsock/peer_map/endpoint.rs [hunk 2]
```diff
@@ -438,7 +438,7 @@ impl Endpoint {
             })
             .collect();
         let ping_needed = !pings.is_empty();
-        let have_endpoints = !self.direct_addr_state.is_empty();
+        let have_alive_endpoints = self.direct_addr_state.values().any(|e| e.is_alive());
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply iroh-net/src/magicsock/peer_map/endpoint.rs:441]: why?

**Review by flub — APPROVED** [review]

**Pull request merged**